src/parquet/ReadParquet.cpp

```diff
@@ -1,5 +1,407 @@
+                                     &values_read);
+        // if values_read is 0, that means that it was a null value
+        if(values_read > 0) {
+          chpl_ptr[row_idx] = (ChplType) value;
+        }
+        else {
+          chpl_ptr[row_idx] = NAN;
+        }
+        row_idx++;
+        num_read++;
+      }
+    }
+    return num_read;
+  }
+
+  // For Parquet's "Decimal" types -- typically larger, or more precise floats
+  template<> int64_t read<arrow::Decimal128Type>() {
+    using Types = TypeBundle<arrow::Decimal128Type>;
+    using ReaderType = typename Types::ReaderType;
+    using ChplType = typename Types::ChplType;
+    using PqType = typename Types::PqType;
+
+    auto chpl_ptr = (ChplType*)chpl_arr;
+    ReaderType* reader = static_cast<ReaderType*>(column_reader.get());
+    startIdx -= reader->Skip(*startIdx);
+
+    // TODO find better variable names. values_read and num_read are confusing
+    int64_t values_read = 0;
+    int64_t num_read = 0;
+
+    using LogType = parquet::DecimalLogicalType;
+    const auto& type = dynamic_cast<const LogType&>(*col_info->logical_type());
+    const int64_t precision = type.precision();
+
+    // In ReadParquet.cpp, there is a basic look up table for this. But number
+    // of required bytes can be found mathematically:
+    int numbits = ceil(precision*3.321928);  // the magic number is the constant
+                                             // from log2(10^precision)
+    if (numbits%8==0) numbits++;             // add a bit for the sign bit if we
+                                             // are at the byte boundary
+    const auto numbytes = ceil(numbits/8.0);
+
+    while (reader->HasNext() && row_idx < numElems) {
+      PqType value;
+      (void)reader->ReadBatch(1, nullptr, nullptr, &value, &values_read);
+      arrow::Decimal128 v;
+      PARQUET_ASSIGN_OR_THROW(v,
+                              ::arrow::Decimal128::FromBigEndian(value.ptr,
+                                                                 numbytes));
+
+      chpl_ptr[row_idx] = v.ToDouble(0);
+      row_idx+=values_read;
+      num_read+=values_read;
+    }
+
+    return num_read;
+  }
+
+  template<typename Types>
+  int64_t _readShortIntegral() {
+    using ReaderType = typename Types::ReaderType;
+    using ChplType = typename Types::ChplType;
+    using PqType = typename Types::PqType;
+
+    auto chpl_ptr = (ChplType*)chpl_arr;
+    ReaderType* reader = static_cast<ReaderType*>(column_reader.get());
+    *startIdx -= reader->Skip(*startIdx);
+
+    // TODO find better variable names. values_read and num_read are confusing
+    int64_t values_read = 0;
+
+    int64_t num_read = 0;
+    if (not hasNonFloatNulls) {
+      // TODO we can read into the actual Chapel array, and then space the data
+      // out in reverse order. I don't think we need to have this temporary
+      // buffer
+      PqType* tmpArr = (PqType*)malloc(batchSize * sizeof(PqType));
```

Comment:
use `new` instead of `malloc`

Reply:
Leaving `malloc` as I want to remove this allocation altogether